File: C/assignment-10/reverse_ll_rec.java
import java.util.*;
class reverse_ll_rec {
    static class Node {
        int data;
        Node next;
        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }
    public static void printList(Node head) {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }
    public static Node reverseLinkedList(Node head) {
        if (head == null || head.next == null) return head;
        Node newHead = reverseLinkedList(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }
    public static void main(String[] args) {
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);
        head.next.next.next = new Node(4);
        System.out.println("Original list:");
        printList(head);
        head = reverseLinkedList(head);
        System.out.println("Reversed list:");
        printList(head);
    }
}

File: C/reverse_ll_rec.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

Node* reverse(Node *head) {

    // Base case
    if (head == NULL || head->next == NULL) {
        return head;
    }

    // Reverse the rest of the list
    Node *newHead = reverse(head->next);

    // Put current node after its next node
    head->next->next = head;

    // Remove old connection
    head->next = NULL;

    return newHead;
}

void printList(Node *head) {
    while (head != NULL) {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

int main() {

    Node *head = malloc(sizeof(Node));
    Node *second = malloc(sizeof(Node));
    Node *third = malloc(sizeof(Node));
    Node *fourth = malloc(sizeof(Node));

    head->data = 1;
    head->next = second;

    second->data = 2;
    second->next = third;

    third->data = 3;
    third->next = fourth;

    fourth->data = 4;
    fourth->next = NULL;

    printf("Original list:\n");
    printList(head);

    head = reverse(head);

    printf("Reversed list:\n");
    printList(head);

    return 0;
}